pubsub-open-telemetry/unary_pull_subscriber.cc

```diff
@@ -0,0 +1,72 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/opentelemetry/configure_basic_tracing.h"
+#include "google/cloud/opentelemetry_options.h"
+#include "google/cloud/pubsub/message.h"
+#include "google/cloud/pubsub/publisher.h"
+#include "google/cloud/pubsub/subscriber.h"
+#include "google/cloud/pubsub/subscription.h"
+#include <iostream>
+
+int main(int argc, char* argv[]) try {
+  if (argc != 4) {
+    std::cerr << "Usage: " << argv[0]
+              << " <project-id> <topic-id> <subscription-id>\n";
+    return 1;
+  }
+
+  std::string const project_id = argv[1];
+  std::string const topic_id = argv[2];
+  std::string const subscription_id = argv[3];
+
+  // Create a few namespace aliases to make the code easier to read.
+  namespace gc = ::google::cloud;
+  namespace otel = gc::otel;
+  namespace pubsub = gc::pubsub;
+
+  auto project = gc::Project(project_id);
+  auto configuration = otel::ConfigureBasicTracing(project);
+
+  // Publish a message with tracing enabled.
+  auto publisher = pubsub::Publisher(pubsub::MakePublisherConnection(
+      pubsub::Topic(project_id, topic_id),
+      gc::Options{}.set<gc::OpenTelemetryTracingOption>(true)));
+  auto id = publisher.Publish(pubsub::MessageBuilder().SetData("Hi!").Build())
+                .then([](gc::future<gc::StatusOr<std::string>> f) {
+                  auto id = f.get();
+                  if (!id) {
+                    std::cout << "Error in publish: " << id.status() << "\n";
+                    return;
+                  }
+                  std::cout << "Sent message with id: (" << *id << ")\n";
+                });
+  // Block until the message is actually sent.
+  id.get();
+
+  // Receive a message using unary pull with tracing enabled.
+  auto subscriber = pubsub::Subscriber(pubsub::MakeSubscriberConnection(
+      pubsub::Subscription(project_id, subscription_id),
+      gc::Options{}.set<gc::OpenTelemetryTracingOption>(true)));
+
+  auto response = subscriber.Pull();
+  if (!response) throw std::move(response).status();
```

Comment:
Since we are throwing on error you could use

```suggestion
  auto response = subscriber.Pull().value();
 ```

Though that throws a different exception (`std::runtime_error`).